rollup/core/classes/RollupQueryBuilder.cls

```diff
@@ -108,7 +109,11 @@ public without sharing class RollupQueryBuilder {
 
         fieldNames.add('TYPEOF ' + whoOrWhat + ' WHEN ' + relationshipName + ' THEN ' + String.join(typeOfFields, ',') + ' END');
 
+        Integer replaceIndex = optionalWhereClause.indexOf(indexer + relationshipName + '\'');
         optionalWhereClause = optionalWhereClause.replace(indexer + relationshipName + '\'', '').replace(whereClause, '');
+        if (optionalWhereClause.substring(replaceIndex).toLowerCase().containsOnly(VERBOTEN_ENDINGS)) {
```

Comment:
Yeah that's a good q! It's because of [RollupFullRecalcTests.shouldAllowGrandparentRollupFromParentWithPolymorphicFields](https://github.com/jamessimone/apex-rollup/blob/7a64e906b76ae8b9a2781e4f7fd4c62f22c427f1/extra-tests/classes/RollupFullRecalcTests.cls#L523).

At a certain point, we end up with a totally cursed where clause like this one: `WHERE Id != :objIds AND (((ParentId != \'someAccountId\' OR Parent.Type = \'Account\') AND Parent.Owner.Id = :recordIds))`

- old: a really ugly reliance on `RollupEvaluator.WhereFieldEvaluator.getSafeQuery()` (which, thanks to your question, I'm realizing I can fully remove) that did post-processing on that where clause after creating the necessary TYPEOF clause
- new (unpushed as of yet): there's no `getSafeQuery()`, which simply ends up bringing the clause that's an issue (`AND ((((ParentId = '0016g00002dlpjQAAQ' OR ) AND )))`) to the forefront when the query is made.

All of that is to say: I think there's still room for improvement on the `RollupQueryBuilder.adjustWhereClauseForPolymorphicFields` method, but this originally came up because I didn't realize the extent to which the query builder and evaluator were coupled, and your comment has allowed me to remove one of those sources of coupling. Hopefully what that means long-term is that this if clause can be removed, but in the meantime at least it's being "cleaned up" in full, guilty scope of the current code, without hidden side-effects elsewhere

Reply:
Ahhhh ok, that makes sense - yeah, I could see revisiting that scenario at some point to see if there's a way to improve/avoid that situation, but I don't think it's a big deal to leave it as-is for now.

Reply:
For sure. At some point (soon, hopefully), I'd like to apply the same scrutiny to `RollupQueryBuilder` that I just did to `RollupEvaluator`, as that class is currently responsible for a lot of cleanup as far as how a "valid query" is constituted, and it would feel really nice to reduce the tech debt here as well